# Explatory Data Analysis on Wide World Importers
**Brayden Parker**

In this analysis we will be exploring all the different relations and connections within the Wide World Importers business, seeing correlations, cleaning data, and making inferences based on the data provided.

### First we will start by importing the data and seeing basic information about it.

In [1]:
import pandas as pd
df = pd.read_csv('invoices.csv')
df.head()

<ipython-input-1-486d913f36dd>:2: DtypeWarning: Columns (12) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('invoices.csv')


InvoiceLineID  InvoiceID  StockItemID  \
0              1          1         67.0   
1              2          2         50.0   
2              3          2         10.0   
3              4          3        114.0   
4              5          4        206.0   

                                         Description  PackageTypeID  Quantity  \
0           Ride on toy sedan car (Black) 1/12 scale              7        10   
1  Developer joke mug - old C developers never di...              7         9   
2               USB food flash drive - chocolate bar              7         9   
3                 Superhero action jacket (Blue) XXL              7         3   
4         Permanent marker black 5mm nib (Black) 5mm              7        96   

   UnitPrice  TaxRate  TaxAmount  LineProfit  ExtendedPrice  LastEditedBy  \
0      230.0       15      345.0       850.0         2645.0             7   
1       13.0       15        NaN        76.5         -999.0             7   
2       32.0       15       43.2       180.0          331.2             7   
3       30.0       15        NaN        24.0         -999.0             7   
4        2.7       15        NaN        96.0            NaN             7   

   LastEditedWhen          Customer  
0             NaN     Stuff by Stew  
1  1/1/2013 12:00   Wholesaler Plus  
2             NaN   Wholesaler Plus  
3             NaN   Big Buys Retail  
4  1/1/2013 12:00  Terry's Trinkets

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230548 entries, 0 to 230547
Data columns (total 14 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   InvoiceLineID   230548 non-null  int64  
 1   InvoiceID       230548 non-null  int64  
 2   StockItemID     230087 non-null  float64
 3   Description     229876 non-null  object 
 4   PackageTypeID   230548 non-null  int64  
 5   Quantity        230548 non-null  int64  
 6   UnitPrice       230548 non-null  float64
 7   TaxRate         230548 non-null  int64  
 8   TaxAmount       109623 non-null  float64
 9   LineProfit      230548 non-null  float64
 10  ExtendedPrice   113081 non-null  float64
 11  LastEditedBy    230548 non-null  int64  
 12  LastEditedWhen  6 non-null       object 
 13  Customer        230548 non-null  object 
dtypes: float64(5), int64(6), object(3)
memory usage: 24.6+ MB


In [3]:
df.describe()

InvoiceLineID      InvoiceID    StockItemID  PackageTypeID  \
count  230548.000000  230548.000000  230087.000000  230548.000000   
mean   114127.066880   35177.386891     110.181827       7.073638   
std     65900.142929   20338.920057      63.722514       0.644528   
min         1.000000       1.000000       1.000000       1.000000   
25%     57062.750000   17570.000000      54.000000       7.000000   
50%    114130.500000   35151.000000     111.000000       7.000000   
75%    171203.250000   52766.000000     165.000000       7.000000   
max    228265.000000   70510.000000     227.000000      10.000000   

            Quantity      UnitPrice        TaxRate      TaxAmount  \
count  230548.000000  230548.000000  230548.000000  109623.000000   
mean       39.252524      45.539361      14.463409     113.328006   
std        55.193934     139.510031       3.904056     221.866696   
min         1.000000       0.660000     -15.000000       0.380000   
25%         5.000000      13.000000      15.000000      14.400000   
50%        10.000000      18.000000      15.000000      34.500000   
75%        60.000000      32.000000      15.000000     129.600000   
max      3150.000000    1899.000000      15.000000   14994.000000   

         LineProfit  ExtendedPrice   LastEditedBy  
count  230548.00000  113081.000000  230548.000000  
mean      375.38894     740.092626      10.800480  
std       753.17041    1705.229089       5.509623  
min      -645.00000    -999.000000       2.000000  
25%        51.00000      89.700000       6.000000  
50%       120.00000     248.400000      11.000000  
75%       390.00000     919.080000      16.000000  
max      9200.00000  114954.000000      20.000000

We can already see a lot about this data set from our prelimanary analyzing. There are 230,548 seperate rows, and 14 columns.

Many columns contain some null values, such as the StockItemID column, and the Description column. Some columns contain a lot of null values such as the ExtendedPrice column and the TaxAmount column as well as the LastEditedWhen column.

Next looking at our data described we see a few logical errors, such as the lowest tax rate being negative, and the lowest extended price being negative as well. Now we can see that the lowest value in line profit is negative, but that could just be because the company lost money on the invoice.

### Now we will begin cleaning the data by fixing standardization errors, and making sure everything follows business logic.

In [4]:
df['Customer'].value_counts()

Customer
Terry's Trinkets       85593
Stuff by Stew          62752
Stuf by Stew           31984
Wholesaler Plus        15735
Big Buys Retail        13461
Terry's Trinket         9483
stuff by stew           7068
Dennis Distribution     4472
Name: count, dtype: int64

In [5]:
df['Customer'] = df['Customer'].str.lower().str.strip(' , ')
df['Customer'] = df['Customer'].str.replace(r'\bterry\'s\ trinket\b', 'terry\'s trinkets', regex=True)
df['Customer'] = df['Customer'].str.replace(r'\bstuf\ by\ stew\b', 'stuff by stew', regex=True)
df['Customer'].unique()

array(['stuff by stew', 'wholesaler plus', 'big buys retail',
       "terry's trinkets", 'dennis distribution'], dtype=object)

In [6]:
df['Customer'].value_counts()

Customer
stuff by stew          101804
terry's trinkets        95076
wholesaler plus         15735
big buys retail         13461
dennis distribution      4472
Name: count, dtype: int64

From standardizing the names of the Customers, we can see that stuff by stew has the most invoices.

Next we will look at the LastEditedWhen column.

In [7]:
df['LastEditedWhen'].isnull().sum()

230542

We can see that there is only 6 values in this column that are not null. 6! That's very few for a dataset with over 200,000 rows. Our best course of action here is to drop the column

In [8]:
df.drop('LastEditedWhen', axis=1, inplace=True)

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230548 entries, 0 to 230547
Data columns (total 13 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   InvoiceLineID  230548 non-null  int64  
 1   InvoiceID      230548 non-null  int64  
 2   StockItemID    230087 non-null  float64
 3   Description    229876 non-null  object 
 4   PackageTypeID  230548 non-null  int64  
 5   Quantity       230548 non-null  int64  
 6   UnitPrice      230548 non-null  float64
 7   TaxRate        230548 non-null  int64  
 8   TaxAmount      109623 non-null  float64
 9   LineProfit     230548 non-null  float64
 10  ExtendedPrice  113081 non-null  float64
 11  LastEditedBy   230548 non-null  int64  
 12  Customer       230548 non-null  object 
dtypes: float64(5), int64(6), object(2)
memory usage: 22.9+ MB


We can now see that the next two columns with the most null values are TaxAmount and ExtendedPice respectively.

We have another problem now with the data, some of the StockItemId and Description columns are both null in some rows, which leaves us without being able to impute data for those rows.

In [10]:
null_id_and_desc = df['StockItemID'].isnull() & df['Description'].isnull()
df['null_id_and_desc'] = null_id_and_desc
df['null_id_and_desc'].sum()

461

In [11]:
df[df['null_id_and_desc'] == True].describe()

InvoiceLineID     InvoiceID  StockItemID  PackageTypeID    Quantity  \
count     461.000000    461.000000          0.0     461.000000  461.000000   
mean   112079.548807  34549.648590          NaN       7.158351   41.876356   
std     68765.368575  21224.946465          NaN       0.703161   55.541797   
min       597.000000    251.000000          NaN       1.000000    1.000000   
25%     51382.000000  15839.000000          NaN       7.000000    5.000000   
50%    113336.000000  34901.000000          NaN       7.000000   10.000000   
75%    174077.000000  53655.000000          NaN       7.000000   60.000000   
max    227910.000000  70407.000000          NaN      10.000000  252.000000   

        UnitPrice     TaxRate    TaxAmount   LineProfit  ExtendedPrice  \
count   461.00000  461.000000   215.000000   461.000000     223.000000   
mean     48.88692   14.533623   130.333209   412.370716     864.816996   
std     162.09547    3.678589   255.183145   800.544226    1991.010590   
min       0.66000  -15.000000     1.950000  -120.000000    -999.000000   
25%      13.00000   15.000000    15.600000    50.000000      92.720000   
50%      18.00000   15.000000    36.000000   135.000000     257.600000   
75%      32.00000   15.000000   148.500000   453.600000    1018.440000   
max    1899.00000   15.000000  2278.800000  7360.000000   17470.800000   

       LastEditedBy  
count    461.000000  
mean      11.080260  
std        5.358461  
min        2.000000  
25%        6.000000  
50%       11.000000  
75%       16.000000  
max       20.000000

In [12]:
df.describe()

InvoiceLineID      InvoiceID    StockItemID  PackageTypeID  \
count  230548.000000  230548.000000  230087.000000  230548.000000   
mean   114127.066880   35177.386891     110.181827       7.073638   
std     65900.142929   20338.920057      63.722514       0.644528   
min         1.000000       1.000000       1.000000       1.000000   
25%     57062.750000   17570.000000      54.000000       7.000000   
50%    114130.500000   35151.000000     111.000000       7.000000   
75%    171203.250000   52766.000000     165.000000       7.000000   
max    228265.000000   70510.000000     227.000000      10.000000   

            Quantity      UnitPrice        TaxRate      TaxAmount  \
count  230548.000000  230548.000000  230548.000000  109623.000000   
mean       39.252524      45.539361      14.463409     113.328006   
std        55.193934     139.510031       3.904056     221.866696   
min         1.000000       0.660000     -15.000000       0.380000   
25%         5.000000      13.000000      15.000000      14.400000   
50%        10.000000      18.000000      15.000000      34.500000   
75%        60.000000      32.000000      15.000000     129.600000   
max      3150.000000    1899.000000      15.000000   14994.000000   

         LineProfit  ExtendedPrice   LastEditedBy  
count  230548.00000  113081.000000  230548.000000  
mean      375.38894     740.092626      10.800480  
std       753.17041    1705.229089       5.509623  
min      -645.00000    -999.000000       2.000000  
25%        51.00000      89.700000       6.000000  
50%       120.00000     248.400000      11.000000  
75%       390.00000     919.080000      16.000000  
max      9200.00000  114954.000000      20.000000

In [13]:
df.drop(df[df['null_id_and_desc'] == True].index).describe()

InvoiceLineID      InvoiceID    StockItemID  PackageTypeID  \
count  230087.000000  230087.000000  230087.000000  230087.000000   
mean   114131.169266   35178.644621     110.181827       7.073468   
std     65894.369187   20337.134754      63.722514       0.644395   
min         1.000000       1.000000       1.000000       1.000000   
25%     57077.500000   17574.000000      54.000000       7.000000   
50%    114131.000000   35151.000000     111.000000       7.000000   
75%    171196.500000   52764.000000     165.000000       7.000000   
max    228265.000000   70510.000000     227.000000      10.000000   

            Quantity      UnitPrice        TaxRate      TaxAmount  \
count  230087.000000  230087.000000  230087.000000  109408.000000   
mean       39.247267      45.532654      14.463268     113.294589   
std        55.193231     139.461436       3.904501     221.796355   
min         1.000000       0.660000     -15.000000       0.380000   
25%         5.000000      13.000000      15.000000      14.400000   
50%        10.000000      18.000000      15.000000      34.500000   
75%        60.000000      32.000000      15.000000     129.600000   
max      3150.000000    1899.000000      15.000000   14994.000000   

          LineProfit  ExtendedPrice   LastEditedBy  
count  230087.000000  112858.000000  230087.000000  
mean      375.314843     739.846178      10.799919  
std       753.072527    1704.618239       5.509919  
min      -645.000000    -999.000000       2.000000  
25%        51.000000      89.700000       6.000000  
50%       120.000000     248.400000      11.000000  
75%       390.000000     919.080000      16.000000  
max      9200.000000  114954.000000      20.000000

As we can see in the above results, so little changes from dropping these rows, we can drop them and preserve our data without consequence.

In [14]:
df.drop(df[df['null_id_and_desc'] == True].index, inplace=True)

Now we will observe the negative values in the TaxRate column and decide how to clean them.

In [15]:
df.describe()

InvoiceLineID      InvoiceID    StockItemID  PackageTypeID  \
count  230087.000000  230087.000000  230087.000000  230087.000000   
mean   114131.169266   35178.644621     110.181827       7.073468   
std     65894.369187   20337.134754      63.722514       0.644395   
min         1.000000       1.000000       1.000000       1.000000   
25%     57077.500000   17574.000000      54.000000       7.000000   
50%    114131.000000   35151.000000     111.000000       7.000000   
75%    171196.500000   52764.000000     165.000000       7.000000   
max    228265.000000   70510.000000     227.000000      10.000000   

            Quantity      UnitPrice        TaxRate      TaxAmount  \
count  230087.000000  230087.000000  230087.000000  109408.000000   
mean       39.247267      45.532654      14.463268     113.294589   
std        55.193231     139.461436       3.904501     221.796355   
min         1.000000       0.660000     -15.000000       0.380000   
25%         5.000000      13.000000      15.000000      14.400000   
50%        10.000000      18.000000      15.000000      34.500000   
75%        60.000000      32.000000      15.000000     129.600000   
max      3150.000000    1899.000000      15.000000   14994.000000   

          LineProfit  ExtendedPrice   LastEditedBy  
count  230087.000000  112858.000000  230087.000000  
mean      375.314843     739.846178      10.799919  
std       753.072527    1704.618239       5.509919  
min      -645.000000    -999.000000       2.000000  
25%        51.000000      89.700000       6.000000  
50%       120.000000     248.400000      11.000000  
75%       390.000000     919.080000      16.000000  
max      9200.000000  114954.000000      20.000000

Seeing as the average TaxRate is so close to being 15, seeing negative 15 doesn't make much sense on a logical sense, so converting it to a postive number makes the most sense here.

In [16]:
df['TaxRate'] = df['TaxRate'].abs()
df.describe()

InvoiceLineID      InvoiceID    StockItemID  PackageTypeID  \
count  230087.000000  230087.000000  230087.000000  230087.000000   
mean   114131.169266   35178.644621     110.181827       7.073468   
std     65894.369187   20337.134754      63.722514       0.644395   
min         1.000000       1.000000       1.000000       1.000000   
25%     57077.500000   17574.000000      54.000000       7.000000   
50%    114131.000000   35151.000000     111.000000       7.000000   
75%    171196.500000   52764.000000     165.000000       7.000000   
max    228265.000000   70510.000000     227.000000      10.000000   

            Quantity      UnitPrice        TaxRate      TaxAmount  \
count  230087.000000  230087.000000  230087.000000  109408.000000   
mean       39.247267      45.532654      14.977248     113.294589   
std        55.193231     139.461436       0.336518     221.796355   
min         1.000000       0.660000      10.000000       0.380000   
25%         5.000000      13.000000      15.000000      14.400000   
50%        10.000000      18.000000      15.000000      34.500000   
75%        60.000000      32.000000      15.000000     129.600000   
max      3150.000000    1899.000000      15.000000   14994.000000   

          LineProfit  ExtendedPrice   LastEditedBy  
count  230087.000000  112858.000000  230087.000000  
mean      375.314843     739.846178      10.799919  
std       753.072527    1704.618239       5.509919  
min      -645.000000    -999.000000       2.000000  
25%        51.000000      89.700000       6.000000  
50%       120.000000     248.400000      11.000000  
75%       390.000000     919.080000      16.000000  
max      9200.000000  114954.000000      20.000000

Now we need to look at all the null values in the TaxAmount column. TaxAmount is easily calculated by multiplying the UnitPrice, Quantity and Tax Rate together and then dividing by 100 to see how much Tax is paid. Seeing how easy this calculation is, we can impute all of the null values in the TaxAmount column.

In [17]:
calculate_tax_amount_quantity_unitprice_taxrate = (df['Quantity'] * df['UnitPrice'] * df['TaxRate']) / 100
df['TaxAmount'] = calculate_tax_amount_quantity_unitprice_taxrate
df.describe()

InvoiceLineID      InvoiceID    StockItemID  PackageTypeID  \
count  230087.000000  230087.000000  230087.000000  230087.000000   
mean   114131.169266   35178.644621     110.181827       7.073468   
std     65894.369187   20337.134754      63.722514       0.644395   
min         1.000000       1.000000       1.000000       1.000000   
25%     57077.500000   17574.000000      54.000000       7.000000   
50%    114131.000000   35151.000000     111.000000       7.000000   
75%    171196.500000   52764.000000     165.000000       7.000000   
max    228265.000000   70510.000000     227.000000      10.000000   

            Quantity      UnitPrice        TaxRate      TaxAmount  \
count  230087.000000  230087.000000  230087.000000  230087.000000   
mean       39.247267      45.532654      14.977248     112.952119   
std        55.193231     139.461436       0.336518     219.486822   
min         1.000000       0.660000      10.000000       0.375000   
25%         5.000000      13.000000      15.000000      14.400000   
50%        10.000000      18.000000      15.000000      34.500000   
75%        60.000000      32.000000      15.000000     129.600000   
max      3150.000000    1899.000000      15.000000   14994.000000   

          LineProfit  ExtendedPrice   LastEditedBy  
count  230087.000000  112858.000000  230087.000000  
mean      375.314843     739.846178      10.799919  
std       753.072527    1704.618239       5.509919  
min      -645.000000    -999.000000       2.000000  
25%        51.000000      89.700000       6.000000  
50%       120.000000     248.400000      11.000000  
75%       390.000000     919.080000      16.000000  
max      9200.000000  114954.000000      20.000000

We can now see that we have no null values in the TaxAmount column, and our mean value changed very little. This was a succesful imputation.

Now lets deal with all our null values in the ExtendedPrice column by imputing the values as well. To calculate the ExtendedPrice we need to multiply UnitPrice and Quantity and the add the TaxAmount. We have all of those values so lets fill in our null values.

In [18]:
calculate_extended_price_quantity_unitprice_taxamount = (df['Quantity'] * df['UnitPrice']) + df['TaxAmount']
df['ExtendedPrice'] = calculate_extended_price_quantity_unitprice_taxamount
df.describe()

InvoiceLineID      InvoiceID    StockItemID  PackageTypeID  \
count  230087.000000  230087.000000  230087.000000  230087.000000   
mean   114131.169266   35178.644621     110.181827       7.073468   
std     65894.369187   20337.134754      63.722514       0.644395   
min         1.000000       1.000000       1.000000       1.000000   
25%     57077.500000   17574.000000      54.000000       7.000000   
50%    114131.000000   35151.000000     111.000000       7.000000   
75%    171196.500000   52764.000000     165.000000       7.000000   
max    228265.000000   70510.000000     227.000000      10.000000   

            Quantity      UnitPrice        TaxRate      TaxAmount  \
count  230087.000000  230087.000000  230087.000000  230087.000000   
mean       39.247267      45.532654      14.977248     112.952119   
std        55.193231     139.461436       0.336518     219.486822   
min         1.000000       0.660000      10.000000       0.375000   
25%         5.000000      13.000000      15.000000      14.400000   
50%        10.000000      18.000000      15.000000      34.500000   
75%        60.000000      32.000000      15.000000     129.600000   
max      3150.000000    1899.000000      15.000000   14994.000000   

          LineProfit  ExtendedPrice   LastEditedBy  
count  230087.000000  230087.000000  230087.000000  
mean      375.314843     867.637267      10.799919  
std       753.072527    1683.168110       5.509919  
min      -645.000000       2.875000       2.000000  
25%        51.000000     110.400000       6.000000  
50%       120.000000     264.500000      11.000000  
75%       390.000000     993.600000      16.000000  
max      9200.000000  114954.000000      20.000000

Things changed here just a little more than the last imputation, but our calculations were done correctly.

Now we are finding outliers in the Quantity and ExtendedPrice columns using a z score threshold.

In [19]:
quantity_z_score = (df['Quantity'] - df['Quantity'].mean()) / df['Quantity'].std()
extendedprice_z_score = (df['ExtendedPrice'] - df['ExtendedPrice'].mean()) / df['ExtendedPrice'].std()

In [20]:
outliers_quantity = df[quantity_z_score.abs() > 3]
outliers_extendedprice = df[extendedprice_z_score.abs() > 3]

In [21]:
outliers_quantity.count()

InvoiceLineID       5987
InvoiceID           5987
StockItemID         5987
Description         5983
PackageTypeID       5987
Quantity            5987
UnitPrice           5987
TaxRate             5987
TaxAmount           5987
LineProfit          5987
ExtendedPrice       5987
LastEditedBy        5987
Customer            5987
null_id_and_desc    5987
dtype: int64

In [22]:
outliers_extendedprice.count()

InvoiceLineID       4522
InvoiceID           4522
StockItemID         4522
Description         4517
PackageTypeID       4522
Quantity            4522
UnitPrice           4522
TaxRate             4522
TaxAmount           4522
LineProfit          4522
ExtendedPrice       4522
LastEditedBy        4522
Customer            4522
null_id_and_desc    4522
dtype: int64

We can see that there are 5,987 values in the Quantity column and 4,522 in the ExtendedPrice column that exceed our >3 z-score threshold which can be considered outliers.

Now we need to decide how to find duplicated in our data, because the InvoiceLineID and the InvoiceID can have matching ID's. We will find duplicates off the InvoiceLineID.

In [ ]:
df.drop_duplicates(subset='InvoiceLineID', inplace=True)
df.describe()

InvoiceLineID      InvoiceID    StockItemID  PackageTypeID  \
count  227811.000000  227811.000000  227811.000000  227811.000000   
mean   114138.092138   35180.773211     110.183512       7.073600   
std     65889.140423   20335.528618      63.730057       0.644306   
min         1.000000       1.000000       1.000000       1.000000   
25%     57082.500000   17575.500000      54.000000       7.000000   
50%    114135.000000   35152.000000     111.000000       7.000000   
75%    171192.500000   52762.500000     165.000000       7.000000   
max    228265.000000   70510.000000     227.000000      10.000000   

            Quantity      UnitPrice        TaxRate      TaxAmount  \
count  227811.000000  227811.000000  227811.000000  227811.000000   
mean       39.251366      45.584056      14.977284     113.027735   
std        55.207157     139.811797       0.336252     219.805264   
min         1.000000       0.660000      10.000000       0.375000   
25%         5.000000      13.000000      15.000000      14.400000   
50%        10.000000      18.000000      15.000000      34.500000   
75%        60.000000      32.000000      15.000000     129.600000   
max      3150.000000    1899.000000      15.000000   14994.000000   

          LineProfit  ExtendedPrice   LastEditedBy  
count  227811.000000  227811.000000  227811.000000  
mean      375.497371     868.215594      10.798105  
std       753.950915    1685.608572       5.510464  
min      -645.000000       2.875000       2.000000  
25%        51.000000     110.400000       6.000000  
50%       120.000000     264.500000      11.000000  
75%       390.000000     993.600000      16.000000  
max      9200.000000  114954.000000      20.000000

### Now we are going to answer the 3 questions the executives of the company asked us.
1. Which item has the highest total quantity sold?
2. Which item has generated the most total profit?
3. Which item is purchased in greatest quantities in a single order, on average?

### Question 1

In [24]:
most_sold_item = df.groupby('Description')['Quantity'].sum().sort_values(ascending=False)
most_sold_item.head()

Description
Black and orange fragile despatch tape 48mmx75m     209088
Black and orange fragile despatch tape 48mmx100m    194184
Clear packaging tape 48mmx75m                       159770
3 kg Courier post bag (White) 300x190x95mm          156250
Shipping carton (Brown) 356x356x279mm               152775
Name: Quantity, dtype: int64

### Question 2

In [25]:
highest_profit_item = df.groupby('Description')['LineProfit'].sum().sort_values(ascending=False)
highest_profit_item.head()

Description
20 mm Double sided bubble wrap 50m          5323120.0
Air cushion machine (Blue)                  4437114.0
32 mm Anti static bubble wrap (Blue) 50m    3546700.0
10 mm Anti static bubble wrap (Blue) 50m    3481920.0
32 mm Double sided bubble wrap 50m          2935670.0
Name: LineProfit, dtype: float64

### Question 3

In [28]:
single_order_most_purchased_item = df.groupby('Description')['Quantity'].mean().sort_values(ascending=False)
single_order_most_purchased_item.head()

Description
Black and orange fragile despatch tape 48mmx75m     199.511450
Black and orange fragile despatch tape 48mmx100m    198.755374
Clear packaging tape 48mmx75m                       145.377616
3 kg Courier post bag (White) 300x190x95mm          143.612132
Shipping carton (Brown) 356x356x279mm               141.852368
Name: Quantity, dtype: float64

That is the end of our data analysis. We have cleaned the data, getting rid of columns with too many null values and not enough uses, imputing data into columns, and finding duplicates and outliers and deciding what to do with them.